subprojects/docs/src/docs/userguide/custom_tasks.adoc

```diff
@@ -150,23 +150,30 @@ If you need to use the old API, have a look at the documentation in the link:htt
+To query incremental changes for an input file property, that property always needs to return the same instance.
+The easiest way to accomplish this is to use one of the following types for such properties: `link:{javadocPath}/org/gradle/api/file/RegularFileProperty.html[RegularFileProperty]`, `link:{javadocPath}/org/gradle/api/file/DirectoryProperty.html[DirectoryProperty]` or `link:{javadocPath}/org/gradle/api/file/ConfigurableFileCollection.html[ConfigurableFileCollection]`.
+
+You can learn more about `RegularFileProperty` and `DirectoryProperty` in the <<lazy_configuration#lazy_configuration,Lazy Configuration>> chapter, especially the sections on <<lazy_configuration#sec:lazy_properties,using read-only and configurable properties>> and <<lazy_configuration#sec:working_with_files_in_lazy_properties,lazy file properties>>.
 ====
 
-The incremental task action can then query for file changes for an input file parameter via `link:{groovyDslPath}/org.gradle.work.InputChanges.html#org.gradle.work.InputChanges:getFileChanges(org.gradle.api.file.FileCollection)[InputChanges.getFileChanges]`.
-The method returns an `Iterable` of `link:{javadocPath}/org/gradle/work/FileChange.html[FileChanges]`, which in turn can be queried for
+The incremental task action can use `link:{groovyDslPath}/org.gradle.work.InputChanges.html#org.gradle.work.InputChanges:getFileChanges(org.gradle.api.file.FileCollection)[InputChanges.getFileChanges()]` to find out what files have changed for a given file-based input property, be it of type `RegularFileProperty`, `DirectoryProperty` or `ConfigurableFileCollection`.
+The method returns an `Iterable` of type `link:{javadocPath}/org/gradle/work/FileChange.html[FileChanges]`, which in turn can be queried for the following:
+
+* the link:{javadocPath}/org/gradle/work/FileChange.html#getFile--[affected file]
+* the link:{javadocPath}/org/gradle/work/FileChange.html#getChangeType--[change type] (`ADDED`, `REMOVED` or `MODIFIED`)
+* the link:{javadocPath}/org/gradle/work/FileChange.html#getNormalizedPath--[normalized path] of the changed file
+* the link:{javadocPath}/org/gradle/work/FileChange.html#getFileType--[file type] of the changed file
 
-* the link:{javadocPath}/org/gradle/work/FileChange.html#getFile--[affected file],
-* the link:{javadocPath}/org/gradle/work/FileChange.html#getChangeType--[change type] (`ADDED`, `REMOVED` or `MODIFIED`),
-* the link:{javadocPath}/org/gradle/work/FileChange.html#getNormalizedPath--[normalized path] of the changed file,
-* the link:{javadocPath}/org/gradle/work/FileChange.html#getFileType--[file type] of the changed file.
+The following example demonstrates an incremental task that has a directory input.
```

Comment:
How about: `that has an input directory`?

Reply:
Yes, that works just as well.